common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -138,22 +283,27 @@ public void invoke(@NonNull UsbYubiKeyDevice device) {
                 synchronized (sDeviceLock) {
                     //set device
                     mDevice = device;
+                    //Reset DialogHolder to null if necessary.
+                    //In this case, DialogHolder would be an ErrorDialog if not null.
+                    dialogHolder.dismissDialog();
+                    //Make sure instance variables are reset/set properly and requests are canceled.
                     mDevice.setOnClosed(new Runnable() {
                         @Override
                         public void run() {
                             synchronized (sDeviceLock) {
                                 Logger.verbose(TAG, "A YubiKey device was disconnected");
                                 mDevice = null;
-                                //Call onCancel on current dialog, if dialog is showing.
-                                synchronized (sSmartcardDialogLock) {
-                                    if (mCurrentDialog != null) {
-                                        mCurrentDialog.onCancelCba();
-                                        //Reset dialog to null
-                                        mCurrentDialog = null;
-                                        //TODO: maybe show a dialog that tells user they unplugged too early? I'm not sure this is necessary since I'm thinking in a lot of cases, the user wanted to cancel current flow by unplugging.
-                                        Logger.verbose(TAG, "YubiKey was disconnected while dialog was still displayed.");
-                                    }
+                                //Remove the YKPiv security provider if it was added.
+                                //Note that Security.removeProvider will silently return if YKPiv doesn't exist.
+                                Security.removeProvider("YKPiv");
+                                //Show an error dialog only if a dialog is still showing.
+                                if (dialogHolder.isDialogShowing()) {
+                                    //Call Cancel Cba callback
+                                    dialogHolder.onCancelCba();
+                                    //Show dialog informing users that they have unplugged their device too early.
```

Comment:
How do we know at this layer that it was too early? Is it possible that the flow was actually successfully completed?

Reply:
The current logic is that if any dialog is showing (that includes error dialogs), then we are in the middle of a flow. 
In the case of a successful flow, the pin dialog is first dismissed, and then `proceed` is called on the request object (this is done in the `useSmartcardCertForAuth` method). In this situation, `dialogHolder.isDialogShowing()` should return false, so the cancel callback nor the error dialog should run. 
It should be noted that `setOnClosed` is automatically called when the fragment is destroyed. So I do need to do some more testing/debugging to see the outcome when different exceptions occur, or other errors occur within the `proceed` method.